compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt

```diff
@@ -1,98 +1,177 @@
 import org.jetbrains.kotlin.descriptors.ClassDescriptor
 import org.jetbrains.kotlin.descriptors.ModuleDescriptor
 import org.jetbrains.kotlin.descriptors.PackageViewDescriptor
 import org.jetbrains.kotlin.descriptors.PropertyDescriptor
+import org.jetbrains.kotlin.descriptors.annotations.AnnotationDescriptor
 import org.jetbrains.kotlin.name.FqName
 import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
 import org.jetbrains.kotlin.resolve.scopes.DescriptorKindFilter
-import kotlin.LazyThreadSafetyMode.NONE
 
 internal class ClassScanner {
 
   /**
    * Returns a sequence of contributed classes from the dependency graph. Note that the result
    * includes inner classes already.
    */
+  @Deprecated("Repeatable")
   fun findContributedClasses(
     module: ModuleDescriptor,
     packageName: String,
     annotation: FqName,
     scope: FqName?
   ): Sequence<ClassDescriptor> {
-    val packageDescriptor = module.getPackage(FqName(packageName))
-    return generateSequence(listOf(packageDescriptor)) { subPackages ->
-      subPackages
-        .flatMap { it.subPackages() }
-        .ifEmpty { null }
-    }
-      .flatMap { it.asSequence() }
-      .flatMap {
-        it.memberScope.getContributedDescriptors(DescriptorKindFilter.VALUES)
-          .asSequence()
-      }
-      .filterIsInstance<PropertyDescriptor>()
-      .groupBy { property ->
-        // For each contributed hint there are several properties, e.g. the reference itself
-        // and the scope. Group them by their common name without the suffix.
-        val name = property.name.asString()
-        val suffix = propertySuffixes.firstOrNull { name.endsWith(it) } ?: return@groupBy name
-        name.substringBeforeLast(suffix)
-      }
-      .values
-      .asSequence()
-      .filter { properties ->
-        // Double check that the number of properties matches how many suffixes we have and how
-        // many properties we expect.
-        properties.size == propertySuffixes.size
-      }
-      .map { ContributedHint(it) }
+    return findContributedHints(module, annotation)
       .let { sequence ->
         if (scope != null) {
           sequence
             .filter { hint ->
               // The scope must match what we're looking for.
-              hint.scope.fqNameSafe == scope
+              hint.contributedAnnotationOrNull(scope) != null
             }
         } else {
           sequence
         }
       }
-      .map { hint -> hint.reference }
+      .map { it.descriptor }
       .filter {
         // Check that the annotation really is present. It should always be the case, but it's
         // a safetynet in case the generated properties are out of sync.
-        it.annotationOrNull(annotation, scope) != null
+        it.annotations(annotation, scope).isNotEmpty()
+      }
+  }
+
+  /**
+   * Returns a sequence of contributed hints from the dependency graph. Note that the result
+   * includes inner classes already.
```

Comment:
Mega nit: Note that the result already includes inner classes.

Reply:
Done